vicare/climate.py

```diff
@@ -227,67 +193,47 @@ def set_temperature(self, **kwargs):
-        """Turn away mode off."""
-        self.set_hold_mode(VICARE_HOLD_MODE_OFF)
-        self.schedule_update_ha_state()
+    @property
+    def preset_mode(self):
+        """Return the current preset mode, e.g., home, away, temp."""
+        if self._current_program == VICARE_PROGRAM_COMFORT:
+            return PRESET_COMFORT
+        elif self._current_program == VICARE_PROGRAM_HOLIDAY:
+            return PRESET_AWAY
+        elif self._current_program == VICARE_PROGRAM_ECO:
+            return PRESET_ECO
+        return None
 
-    def set_hold_mode(self, hold_mode):
-        if hold_mode in [VICARE_HOLD_MODE_AWAY, VICARE_HOLD_MODE_HOME]:
-            active_mode = self._api.getActiveMode()
-            self.schedule_update_ha_state()
-            if hold_mode == VICARE_HOLD_MODE_AWAY:
-                success = self._api.setMode(VICARE_MODE_FORCEDREDUCED)
-            else:
-                hold_mode == VICARE_HOLD_MODE_HOME
-                success = self._api.setMode(VICARE_MODE_FORCEDNORMAL)
-            # PyViCare currently returns 'None' on no error, checking for both for future changes
-            if success["error"] is None or success["error"] == 'None':
-                self._hold = hold_mode
-                if active_mode in [VICARE_MODE_OFF, VICARE_MODE_DHW, VICARE_MODE_DHWANDHEATING, "active"]:
-                    self._pre_hold = active_mode
-            else:
-                _LOGGER.error("Failed to set hold mode on ViCare API to %s with status code %s and error %s", hold_mode, success["statusCode"], success["error"])
-        elif hold_mode == VICARE_HOLD_MODE_OFF:
-            if self._pre_hold is not None:
-                success = self._api.setMode(self._pre_hold)
-            else:
-                _LOGGER.info("No stored pre hold mode found - switching to hot water and heating")
-                success = self._api.setMode(VICARE_MODE_DHWANDHEATING)
-            # PyViCare currently returns 'None' on no error, checking for both for future changes
-            if not (success["error"] is None or success["error"] == 'None'):
-                _LOGGER.error("Failed to restore from hold mode on ViCare API to %s with status code %s and error %s", self._pre_hold, success["statusCode"], success["error"])
-        else:
-            _LOGGER.error("Unknown hold mode %s set - ignoring", hold_mode)
+    @property
+    def preset_modes(self):
+        return [PRESET_COMFORT, PRESET_AWAY, PRESET_ECO]
 
-    def turn_on(self):
-        """Turn on."""
-        self._on = True
-        self.schedule_update_ha_state()
-        self._api.setMode(VICARE_MODE_DHWANDHEATING)
+    def set_preset_mode(self, preset_mode):
+        """ 1st deactivate any existing program"""
+        self._api.deactivateProgram(self._current_program)
 
-    def turn_off(self):
-        """Turn off."""
-        self._on = False
-        self.schedule_update_ha_state()
-        self._api.setMode(VICARE_MODE_OFF)
+        """ 2nd: set new mode """
```

Comment:
If I remember correctly, setting programs only works if the current VICARE_MODE = VICARE_MODE_DHWANDHEATING. Assuming the thermostat is already in this mode, will cause problems.

I'm not too sure about how the HASS climate preset are used, or how they should behave, but the presets could be mapped to VICARE_MODES, as far as I'm concerned... The Viessmann programs are just bloat added to this:
PRESET_COMFORT = VICARE_MODE_FORCEDNORMAL
PRESET_ECO = VICARE_MODE_FORCEDREDUCED
PRESET_AWAY = redundant, for me this is the same as ECO

I don't think the Viessmann 'programs' have any added value in this case, but TBD.

Reply:
I believe presets are supposed to be more like VICARE_PROGRAMS not VICARE_MODES.
If you go through the Climate 1.0 refactoring pull request of HA it is quite obvious. There is often a preset for at least eco or comfort.
Let's keep it simple for now and add these two presets. Let's ignore holiday, away, or any other ones for now. and discuss about it in additional PR. Main focus should be to get the initial functionality working